### PR TITLE
Add support for kudo write metrics

### PR DESCRIPTION
**Pull request by liurenjie1024**
This pr address two things:
1. Refactor metrics in gpu exchange to use constants rather string literals.
2. Add support for kudo write metrics. This part relys on https://github.com/NVIDIA/spark-rapids-jni/pull/2630 to be merged.

### CODE REVIEW

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala [hunk 1]
```diff
@@ -276,6 +265,58 @@ abstract class GpuShuffleExchangeExecBase(
 }
 
 object GpuShuffleExchangeExecBase {
+  val METRIC_DATA_SIZE = "dataSize"
+  val METRIC_DESC_DATA_SIZE = "data size"
+  val METRIC_DATA_READ_SIZE = "dataReadSize"
+  val METRIC_DESC_DATA_READ_SIZE = "data read size"
+  val METRIC_SHUFFLE_SERIALIZATION_TIME = "rapidsShuffleSerializationTime"
+  val METRIC_DESC_SHUFFLE_SERIALIZATION_TIME = "rs. serialization time"
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala:273]: Not new with this PR, but "rs." is not a very helpful description and is more cryptic than the metric name itself.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala [hunk 2]
```diff
@@ -315,8 +356,11 @@ object GpuShuffleExchangeExecBase {
       rdd
     }
     val partitioner: GpuExpression = getPartitioner(newRdd, outputAttributes, newPartitioning)
+    val partitonTime: GpuMetric = metrics(METRIC_SHUFFLE_PARTITION_TIME)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala:359]:
```suggestion
    val partitionTime: GpuMetric = metrics(METRIC_SHUFFLE_PARTITION_TIME)
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala [hunk 3]
```diff
@@ -37,6 +38,7 @@ import org.apache.spark.sql.execution.exchange.{Exchange, ShuffleExchangeExec}
 import org.apache.spark.sql.execution.metric._
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.rapids.GpuShuffleDependency
+import org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.{createAdditionalExchangeMetris, METRIC_DATA_READ_SIZE, METRIC_DATA_SIZE, METRIC_DESC_DATA_READ_SIZE, METRIC_DESC_DATA_SIZE, METRIC_DESC_SHUFFLE_COMBINE_TIME, METRIC_DESC_SHUFFLE_DESERIALIZATION_TIME, METRIC_DESC_SHUFFLE_PARTITION_TIME, METRIC_DESC_SHUFFLE_READ_TIME, METRIC_DESC_SHUFFLE_SER_CALC_HEADER_TIME, METRIC_DESC_SHUFFLE_SER_COPY_BUFFER_TIME, METRIC_DESC_SHUFFLE_SER_COPY_HEADER_TIME, METRIC_DESC_SHUFFLE_SERIALIZATION_TIME, METRIC_DESC_SHUFFLE_WRITE_IO_TIME, METRIC_DESC_SHUFFLE_WRITE_TIME, METRIC_SHUFFLE_COMBINE_TIME, METRIC_SHUFFLE_DESERIALIZATION_TIME, METRIC_SHUFFLE_PARTITION_TIME, METRIC_SHUFFLE_READ_TIME, METRIC_SHUFFLE_SER_CALC_HEADER_TIME, METRIC_SHUFFLE_SER_COPY_BUFFER_TIME, METRIC_SHUFFLE_SER_COPY_HEADER_TIME, METRIC_SHUFFLE_SERIALIZATION_TIME, METRIC_SHUFFLE_WRITE_IO_TIME, METRIC_SHUFFLE_WRITE_TIME}
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala:41]:
```suggestion
import org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.createAdditionalExchangeMetrics
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala [hunk 4]
```diff
@@ -276,6 +265,58 @@ abstract class GpuShuffleExchangeExecBase(
...
+  val METRIC_SHUFFLE_READ_TIME = "rapidsShuffleReadTime"
+  val METRIC_DESC_SHUFFLE_READ_TIME = "rs. shuffle read time"
+  val METRIC_SHUFFLE_SER_CALC_HEADER_TIME = "rapidsShuffleSerializationCalcHeaderTime"
+  val METRIC_DESC_SHUFFLE_SER_CALC_HEADER_TIME = "rs. serialization calc header time"
+  val METRIC_SHUFFLE_SER_COPY_HEADER_TIME = "rapidsShuffleSerializationCopyHeaderTime"
+  val METRIC_DESC_SHUFFLE_SER_COPY_HEADER_TIME = "rs. serialization copy header time"
+  val METRIC_SHUFFLE_SER_COPY_BUFFER_TIME = "rapidsShuffleSerializationCopyBufferTime"
+  val METRIC_DESC_SHUFFLE_SER_COPY_BUFFER_TIME = "rs. serialization copy buffer time"
+
+  def createAdditionalExchangeMetris(gpu: GpuExec): Map[String, GpuMetric] = Map(
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala:293]:
```suggestion
  def createAdditionalExchangeMetrics(gpu: GpuExec): Map[String, GpuMetric] = Map(
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala [hunk 5]
```diff
@@ -679,7 +675,6 @@ abstract class RapidsShuffleThreadedReaderBase[K, C](
             }
             val res = currentIter.next()
             val fetchTime = System.nanoTime() - fetchTimeStart
-            deserializationTimeNs.foreach(_ += (fetchTime - readBlockedTime))
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala:682]: I'm not sure the new deserialization time metric does the same thing the old metric did.  The new metric is timing how long it takes the serializer to produce a new batch, but that's measuring I/O time as well as deserialization time once the I/O is completed.  The original metric appears to go out of its way to remove the time spent on I/O to get a more accurate deserialization metric.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala [hunk 6]
```diff
@@ -270,27 +270,32 @@ object GpuShuffleExchangeExecBase {
   val METRIC_DATA_READ_SIZE = "dataReadSize"
   val METRIC_DESC_DATA_READ_SIZE = "data read size"
   val METRIC_SHUFFLE_SERIALIZATION_TIME = "rapidsShuffleSerializationTime"
-  val METRIC_DESC_SHUFFLE_SERIALIZATION_TIME = "rs. serialization time"
+  val METRIC_DESC_SHUFFLE_SERIALIZATION_TIME = "rapids shuffle serialization time"
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala:273]: RAPIDS should always be capitalized in documentation

**Review by revans2 — APPROVED** [review]
The changes look good to me, but I also want to hear from Jason if he has more concerns

**Pull request merged**